python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -186,11 +204,46 @@ def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
             )
         }
 
-        return list(
+        root_partitions = set(
             self.target_subset.filter_asset_keys(
                 assets_with_no_parents_in_target_subset
             ).iterate_asset_partitions()
         )
+        root_and_downstream_partitions = _get_self_and_downstream_targeted_subset(root_partitions)
+        previous_root_and_downstream_partitions = None
+
+        while (
+            root_and_downstream_partitions != self.target_subset
+            and root_and_downstream_partitions
+            != previous_root_and_downstream_partitions  # Check against previous iteration result to exit if no new partitions are targeted
+        ):
+            unreachable_targets = self.target_subset - root_and_downstream_partitions
+
+            unreachable_target_root_partitions = set(
+                unreachable_targets.filter_asset_keys(
+                    AssetSelection.keys(*unreachable_targets.asset_keys)
+                    .sources()
+                    .resolve(unreachable_targets.asset_graph)
+                ).iterate_asset_partitions()
+            )
+
+            root_partitions = root_partitions | unreachable_target_root_partitions
+
+            previous_root_and_downstream_partitions = root_and_downstream_partitions
+            root_and_downstream_partitions = (
+                root_and_downstream_partitions
+                | _get_self_and_downstream_targeted_subset(unreachable_target_root_partitions)
+            )
+
+        if root_and_downstream_partitions == previous_root_and_downstream_partitions:
+            raise DagsterInvariantViolationError(
+                "Unable to determine root partitions for backfill. The following asset partitions"
```

Comment:
Raises an error when certain partitions will never be targeted, so that we can detect this early on

Reply:
Is my understanding correct that it's theoretically impossible to hit this case and this is basically just to prevent infinite loops in case something weird happens?

Reply:
yeah that's right, assuming the logic in this function is correct the user should never hit this.

I added it because it would be quite awful if the daemon hung forever on this infinite loop